libdnf/repo/solvable/Dependency.cpp

```diff
@@ -0,0 +1,157 @@
+#include "Dependency.hpp"
+#include "libdnf/utils/utils.hpp"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+#   include <solv/pool_parserpmrichdep.h>
+#ifdef __cplusplus
+}
+#endif
+
+
+#define NUMBER_OF_ITEMS_IN_REL_DEPENDENCY 3
+
+class Dependency::Impl {
+public:
+    Impl();
+    Impl(const Impl &source);
+    ~Impl() = default;
+private:
+    friend class Dependency;
+
+    void parseAndCreateDependency(const std::string &dependency);
+    void createRelationalDependency(const std::vector<std::string> &results);
+    bool isRichDependency(const std::string &dependency) const;
+    int determineComparisonType(const std::string &type);
+    Pool *getPool();
+
+    Id libsolvableId;
+    DnfSack *sack;
+    int type;
+};
+
+Dependency::Impl::Impl()
+        : libsolvableId(-1)
+        , sack(nullptr)
+        , type(-1)
+{
+}
+
+Dependency::Impl::Impl(const Impl &source)
+{
+    libsolvableId  = source.libsolvableId;
+    sack = source.sack;
+    type = source.type;
+}
+
+void Dependency::Impl::parseAndCreateDependency(const std::string &dependency)
+{
+    std::vector<std::string> results = string::split(dependency, " ");
+
+    if (isRichDependency(dependency)) {
+        libsolvableId = pool_parserpmrichdep(getPool(), dependency.c_str());
+    } else if (results.size() == NUMBER_OF_ITEMS_IN_REL_DEPENDENCY) { // hasRelationalOperator
+        createRelationalDependency(results);
+    } else {
+        libsolvableId = pool_str2id(getPool(), results[0].c_str(), 1 /* create */);
+    }
+}
+
+void Dependency::Impl::createRelationalDependency(const std::vector<std::string> &results)
+{
+    Id id = pool_str2id(getPool(), results[0].c_str(), 1);
+
+    if (!results[2].empty()) {
+        Id evrId = pool_str2id(getPool(), results[2].c_str(), 1);
+        libsolvableId = pool_rel2id(getPool(), id, evrId, determineComparisonType(results[1]), 1);
+    }
+}
+
+bool Dependency::Impl::isRichDependency(const std::string &dependency) const
+{
+    return dependency.find('(', 0) != std::string::npos;
+}
+
+int Dependency::Impl::determineComparisonType(const std::string &type)
+{
+    if (type == "<=") {
+        return REL_LT | REL_EQ;
+    } else if (type == ">=") {
+        return REL_GT | REL_EQ;
+    } else if (type == "==") {
+        return REL_EQ;
+    } else if (type == ">") {
+        return REL_GT;
+    } else if (type == "<") {
+        return REL_LT;
+    }
+
+    throw "Unknown comparison: " + type;
+}
+
+Pool *Dependency::Impl::getPool()
+{
+    return dnf_sack_get_pool(sack);
+}
+
+Dependency::Dependency(DnfSack *sack, int id)
+        : pImpl(new Impl())
+{
+    pImpl->sack = sack;
+    pImpl->libsolvableId = id;
+    pImpl->type = -1;
```

Comment:
I think that the type is not required in general. It is only used for creation of new solvable form string, therefore it can be kept externally.

Reply:
`type` is required. See how dependencies are being put into solvable.

Reply:
Yes, but the information can be passed to libsolv by other way. Like a separate set for each type.

Reply:
Do you know how many different types of dependencies are there? This is not python, you can't have bazillion input arguments.